python/ray/data/datasource/file_based_datasource.py

```diff
@@ -429,6 +318,138 @@ def file_extension_filter(cls) -> Optional[PathPartitionFilter]:
+        import numpy as np
+
+        open_stream_args = self._open_stream_args
+        reader_args = self._reader_args
+        _block_udf = self._block_udf
+
+        paths, file_sizes = self._paths, self._file_sizes
+        if self._partition_filter is not None:
+            paths = self._partition_filter(paths)
+
+        read_stream = self._delegate._read_stream
+        filesystem = _wrap_s3_serialization_workaround(self._filesystem)
+        read_options = reader_args.get("read_options")
+        if read_options is not None:
+            import pyarrow.json as pajson
+
+            if isinstance(read_options, pajson.ReadOptions):
+                _register_arrow_json_readoptions_serializer()
+
+        if open_stream_args is None:
+            open_stream_args = {}
+
+        def read_files(
+            read_paths: List[str],
+            fs: Union["pyarrow.fs.FileSystem", _S3FileSystemWrapper],
+        ) -> Iterable[Block]:
+            logger.debug(f"Reading {len(read_paths)} files.")
+            if isinstance(fs, _S3FileSystemWrapper):
+                fs = fs.unwrap()
+            ctx = DatasetContext.get_current()
+            output_buffer = BlockOutputBuffer(
+                block_udf=_block_udf, target_max_block_size=ctx.target_max_block_size
+            )
+            for read_path in read_paths:
+                compression = open_stream_args.pop("compression", None)
+                if compression is None:
+                    import pyarrow as pa
+
+                    try:
+                        # If no compression manually given, try to detect
+                        # compression codec from path.
+                        compression = pa.Codec.detect(read_path).name
+                    except (ValueError, TypeError):
+                        # Arrow's compression inference on the file path
+                        # doesn't work for Snappy, so we double-check ourselves.
+                        import pathlib
+
+                        suffix = pathlib.Path(read_path).suffix
+                        if suffix and suffix[1:] == "snappy":
+                            compression = "snappy"
+                        else:
+                            compression = None
+                if compression == "snappy":
+                    # Pass Snappy compression as a reader arg, so datasource subclasses
+                    # can manually handle streaming decompression in
+                    # self._delegate._read_stream().
+                    reader_args["compression"] = compression
+                    reader_args["filesystem"] = fs
+                elif compression is not None:
+                    # Non-Snappy compression, pass as open_input_stream() arg so Arrow
+                    # can take care of streaming decompression for us.
+                    open_stream_args["compression"] = compression
+                with self._delegate._open_input_source(
```

Comment:
We shouldn't capture `self` in this closure, since that's going to create serialization issues, e.g. the `self.filesystem` deserialization might break since it's not using our wrapper. We should change `FileBasedDatasource._open_input_source()` to a static method (it's not currently using the instance anyway) and close over that function pointer instead of `self`.